src/hooks/prisma.ts

```diff
@@ -0,0 +1,137 @@
+
+export default function prismaHook(mod: typeof prisma) {
+  // Imported PrismaClient type does not have _request method in type definition.
+  // But we have it in runtime.
+  assert("PrismaClient" in mod);
+  assert(mod.PrismaClient != null);
+  const PC = mod.PrismaClient as { prototype: unknown };
+  const proto = PC.prototype;
+  assert(proto != null && typeof proto === "object");
+  assert("_request" in proto);
+  proto._request = createProxy(proto._request as (...args: unknown[]) => unknown);
+  return mod;
+}
+
+prismaHook.applicable = function (id: string) {
+  return ["@prisma/client"].includes(id);
+};
+
+// https://github.com/prisma/prisma/blob/095cba1a1b79d0d950246b07c9fb48d22fd7f229/packages/client/src/runtime/getPrismaClient.ts#L181
+interface QueryEvent {
+  timestamp: Date;
+  query: string;
+  params: string;
+  duration: number;
+  target: string;
+}
+
+interface PrismaRequestParamsArgs {
+  data?: unknown;
+  include?: unknown;
+  where?: unknown;
+}
+
+interface PrismaRequestParams {
+  action?: string;
+  model?: string;
+  args?: PrismaRequestParamsArgs;
+}
+
+let hookAttached = false;
+
+const functionInfos = new Map<string, FunctionInfo>();
+
+function getFunctionInfo(model: string, action: string, params: ESTree.Parameter[]) {
+  const key = model + "." + action;
+  if (!functionInfos.has(key)) {
+    const info: FunctionInfo = {
+      async: true,
+      generator: false,
+      id: action,
+      params: params,
+      location: { path: "unknown", lineno: 0 },
+      klassOrFile: model,
+      static: true,
+    };
+    functionInfos.set(key, info);
+  }
+
+  return functionInfos.get(key)!;
+}
+
+function createProxy<T extends (...args: unknown[]) => unknown>(prismaClientMethod: T) {
+  return new Proxy(prismaClientMethod, {
+    apply(target, thisArg: unknown, argArray: Parameters<T>) {
+      if (!hookAttached) {
+        hookAttached = true;
+        assert(
+          thisArg != null &&
+            typeof thisArg === "object" &&
+            "_engine" in thisArg &&
+            thisArg._engine != null &&
+            typeof thisArg._engine === "object" &&
+            "config" in thisArg._engine &&
+            thisArg._engine.config != null &&
+            typeof thisArg._engine.config === "object" &&
+            "logLevel" in thisArg._engine.config &&
+            "logQueries" in thisArg._engine.config &&
+            "activeProvider" in thisArg._engine.config &&
+            typeof thisArg._engine.config.activeProvider == "string",
+        );
+
+        const dbType = thisArg._engine.config.activeProvider;
+        thisArg._engine.config.logLevel = "query";
+        thisArg._engine.config.logQueries = true;
+        assert("$on" in thisArg && typeof thisArg.$on === "function");
+        thisArg.$on("query", (queryEvent: QueryEvent) => {
+          const call = recording.sqlQuery(dbType, queryEvent.query);
+          recording.functionReturn(call.id, undefined, queryEvent.duration);
```

Comment:
I'd double check the unit of time here. I tried a request which issued three queries taking 2ms, 0ms, 1ms according to the logs. The AppMap states the duration was 2000ms, 0ms, 1000ms.